bip78/src/receiver/mod.rs

```diff
@@ -37,28 +60,111 @@ impl UncheckedProposal {
+    /// Such so called "non-interactive" receivers are otherwise vulnerable to probing attacks.
+    /// If a sender can make requests at will, they can learn which bitcoin the receiver owns at no cost.
+    /// Broadcasting the Original PSBT after some time in the failure case makes incurs sender cost and prevents probing.
+    ///
+    /// Call this after checking downstream.
+    pub fn assume_tested_and_scheduled_broadcast(self) -> MaybeInputsOwned {
+        MaybeInputsOwned {
             psbt: self.psbt,
+            params: self.params,
         }
     }
 
-    pub fn this_is_purely_interactive_wallet(self) -> UnlockedProposal {
-        UnlockedProposal {
+    /// Call this method if the only way to initiate a PayJoin with this receiver
+    /// requires manual intervention, as in most consumer wallets.
+    ///
+    /// So-called "non-interactive" receivers, like payment processors, that allow arbitrary requests are otherwise vulnerable to probing attacks.
+    /// Those receivers call `get_transaction_to_check_broadcast()` and `attest_tested_and_scheduled_broadcast()` after making those checks downstream.
+    pub fn assume_interactive_receive_endpoint(self) -> MaybeInputsOwned {
+        MaybeInputsOwned {
             psbt: self.psbt,
+            params: self.params,
         }
     }
 }
 
+impl MaybeInputsOwned {
+    /// The receiver should not be able to sign for any of these Original PSBT inputs.
+    /// Check that none of them are owned by the receiver downstream before proceeding.
+    pub fn iter_input_script_pubkeys(&self) -> Vec<Result<&Script, RequestError>> {
+        todo!() // return impl '_ + Iterator<Item = Result<&Script, RequestError>>
+    }
+
+    /// If the sender included inputs that the receiver could sign for in the original PSBT,
+    /// the receiver must either return error original-psbt-rejected or make sure they do not sign those inputs in the payjoin proposal.
+    ///
+    /// Call this after checking downstream.
+    pub fn assume_inputs_not_owned(self) -> MaybeMixedInputScripts {
+        MaybeMixedInputScripts {
+            psbt: self.psbt,
+            params: self.params,
+        }
+    }
+}
+
+impl MaybeMixedInputScripts {
+    /// If there is only 1 input type, the receiver should be able to produce the same
+    /// type.
+    ///
+    /// Check downstream before proceeding.
+    pub fn iter_input_script_types(&self) -> Vec<Result<&AddressType, RequestError>> {
+        todo!() // return Iterator<Item = Result<&AddressType, RequestError>>
+    }
+
+    /// Verify the original transaction did not have mixed input types
+    /// Call this after checking downstream.
+    ///
+    /// Note: mixed spends do not necessarily indicate distinct wallet fingerprints.
+    /// This check is intended to prevent some types of wallet fingerprinting.
+    pub fn assume_no_mixed_input_scripts(self) -> MaybeInputsSeen {
```

Comment:
We can actually do this check internally, no need to burden the consumer with it. The consumer only needs to get one address type which is the same for all inputs and check that one.

Reply:
Is this true? Can we stay up to date on all the possible bitcoin scripts?

Reply:
It's a good point that allowing people to manually check would allow them to work around the library not being updated but if we fail to maintain it they probably should fork it rather than using an obsolete library for potentially security-critical software.

Reply:
it'd be better if the alpha version of this forgoes mixed input script check since we want to use it for loptos. Is it blocking this PR or may we leave it for the next iteration like some of the warns? I'd love for us to get loptos out ASAP

Reply:
Even if we skip the check AFAIK there are no wallets currently that support mixed inputs, so that's still useless. If you know of any wallet that does please let me know.